wolfCLU/src/x509/clu_request_setup.c

```diff
@@ -0,0 +1,292 @@
+/* clu_request_setup.c
+ *
+ * Copyright (C) 2006-2020 wolfSSL Inc.
+ *
+ * This file is part of wolfSSL. (formerly known as CyaSSL)
+ *
+ * wolfSSL is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfSSL is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#include <wolfclu/clu_header_main.h>
+#include <wolfclu/clu_optargs.h>
+#include <wolfclu/x509/clu_request.h>
+#include <wolfclu/x509/clu_cert.h>
+#include <wolfclu/certgen/clu_certgen.h>
+
+int wolfCLU_requestSetup(int argc, char** argv)
+{
+#ifndef WOLFSSL_CERT_REQ
+    printf("wolfSSL not compiled with --enable-certreq\n");
+    return -1;
+#else
+    WOLFSSL_BIO *bioOut = NULL;
+    WOLFSSL_BIO *bioIn  = NULL;
+    WOLFSSL_X509 *x509  = NULL;
+    const WOLFSSL_EVP_MD *md  = NULL;
+    WOLFSSL_EVP_PKEY *pkey = NULL;
+
+    int     ret = 0;
+    int     i   = 0;
+    char*   in  = NULL;
+    char*   out = NULL;
+    char*   config = NULL;
+
+    char*   alg;                /* algorithm being used */
+    int     algCheck =   0;     /* algorithm type */
+    int     oid      =   0;
+    int     outForm = PEM_FORM; /* default to PEM format */
+    int     inForm;
+    int     option;
+    int     long_index = 1;
+    int     days = 0;
+    int     genX509 = 0;
+
+
+    opterr = 0; /* do not display unrecognized options */
+    optind = 0; /* start at indent 0 */
+    while ((option = getopt_long_only(argc, argv, "", req_options,
+                    &long_index )) != -1) {
+
+        switch (option) {
+            case WOLFCLU_INFILE:
+            case WOLFCLU_KEY:
+                in = optarg;
+                bioIn = wolfSSL_BIO_new_file(optarg, "rb");
```

Comment:
Any particular reason to use BIO's instead of native API? Could be `wc_PemPubKeyToDer` (although we are lacking a private key equivalent).

Reply:
BIO's handle nicely if printing out to stdout or if printing to a file or if reading a string from the command line. I think it makes the code cleaner if the operation itself does not need to worry about what type of IO is being used.

Reply:
That's fair enough. BIO's are, in my view, a very good and stable part of the compatibility layer. The `WOLFSSL_X509` handling is not as consistent unfortunately so in future reviews I will be looking for native alternatives to using that part of the compatibility layer.